# biomed-multi-omic

Biomedical foundational models for omics data. This package supports the development of foundation models for scRNA or for DNA data.

### Highlights
- ðŸ§¬ A single package for DNA and RNA Foundation models
- ðŸš€ Built around anndata, HuggingFace transformers and PyTorchLighting to integrate into exsisting software stacks
- ðŸ“ˆ Increased performance on AvgBio scores using Whole Cell Expression Decoder (WCED)
- ðŸ”¬ Most expensive package for Trancriptomic Foundation Models (TPM) and Genomics Foundation Models

## Contents

1. [Introduction](#1-introduction)
2. [Installation](#2-installation)
3. [Quick start](#3-quick-start)
4. [Contributing](#4-contributing)
5. [Citation](#5-citation)

## Introduction

`biomed-multi-omic` is a comprehensive software package that not only facilitates this combinatorial exploration but is also inherently flexible and easily extensible to incorporate novel methods as the field continues to advance. It also facilitates downstream tasks such as cell type annotation, perturbation prediction, and batch effect
correction on benchmark datasets. `biomed-multi-omic` is built upon anndata, HuggingFace Transformers and PyTorchLighting to ensure integration into exsisting software stacks.


## Package Architecture

### RNA Modules

`bmfm-rna` framework diagram schematic shows the modules avaliable for building  Transcriptomics Foundation Model (TPM). A novel contribution of our work is the Whole Cell Expression Decoder (WCED), an innovative pretraining method aimed at improving transcriptomic foundation models. In WCED, the modelâ€™s objective is to reconstruct a full cell expression profile from a partial input's `[CLS]` token representation generated by the transformer encoder. By training models to autocomplete the expresion profiles, WCED improves the modelâ€™s understanding ofunderlying biological processes, resulting in better generalization and more accurate predictions for downstream tasks.

- (A) scRNA data can be read from TileDB-SOMA (eg CellXGene) or h5ad,and processed with litdata to support multi-gpu training.
- (B) Quality control pipeline built on top of Scanpy.
- (C) Data selection techniques to prioritize informative samples.
- (D) Sample transformations/augmentations including read-depth aware downsampling, non-expressed gene sampling, expression binning, gene ordering and dropout.
- (E) Multitask input preparation supports masked language modeling on gene and/or expression inputs, sequence classification across multiple label types (e.g., cell type), joint training of both tasks, and sequence labeling for perturbed expression prediction.
- (F) Model input representation supports integration of external gene-level knowledge, with expression values represented as either discrete tokens or continuous embeddings
- (G) Customized Transformer architectures based on
HuggingFace including BERT, ModernBERT, Performer and Nystromformer to support multi-field inputs (genes, expressions, perturbation state) and multi-task learning.
- (H) Training is implemented with PyTorch Lightning and fully configurable via Hydra, supporting flexible benchmarking. Multi-task training uses weighted losses built with
TorchMetrics, including cross-entropy, focal loss, and zero-inflated MSE. Gradient reversal is incorporated to mitigate batch effects.
- (I) ClearML is used for seamless experiment tracking and interactive visualization of results.
- (J) Captum is used for model interpretability, allowing gene-level attributions to highlight important features driving predictions.

NB - Orange indicates in-development.

![bmfm_omics_workflow](docs/images/package_diagram.png)

## DNA Modules

TBC.


## Installation

We recommend using [uv](https://github.com/astral-sh/uv) to create your enviroment due to it's 10-100x speed up over pip. 

NB - if you do not have `curl` installed you will need to install it before trying to install the tool (e.g. `apt`, `brew` or `choco`). For more information please refer to `curl`'s github:

Install uisng `uv` (recommended):
```sh
python -m venv .venv
source .venv/bin/activate
pip install uv
uv pip install "git+github.com/BiomedSciAI/biomed-multi-omic.git"
```

Install uisng `pip`:
```sh
python -m venv .venv
source .venv/bin/activate
pip install "git+github.com/BiomedSciAI/biomed-multi-omic.git"
```

Install using cloned repo:
```sh
git clone git@github.com:BiomedSciAI/biomed-multi-omic.git
cd biomed-multi-omics
uv venv .venv -p3.12
source ./.venv/bin/activate
uv pip install -e .
```

### Optional dependencies
In addition to the base package there are additional optional dependencies which extends `bmfm-mulit-omics` capabilites further. These include:
- `bulk_rna`: Extends modules for extracting and preprocessing bulk RNA-seq data
- `benchmarking`: Installs additional models used benchmark `bmfm-mulit-omics` against. These include scib, scib-metrics, pyliger, scanorama and harmony-pytorch.
- `test`: Unittest suite which is recommended for development use

To install optional dependencies from this GitHub repository you can use the following structure:

```sh
uv pip install "git+github.com/BiomedSciAI/biomed-multi-omic.git#egg=package[bulk_rna,benchmarking,test,notebook]"
```

## Quick start

### Downloading checkpoint

The model's weights can be aquired from [IBM's HuggingFace collection](https://huggingface.co/ibm-research). The following scRNA models are avaliable:

- MLM+RDA [ibm-research/biomed.rna.bert.110m.mlm.rda.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.mlm.rda.v1)
- MLM+Multitask [ibm-research/biomed.rna.bert.110m.mlm.multitask.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.mlm.multitask.v1)
- WCED+Multitask [ibm-research/biomed.rna.bert.110m.wced.multitask.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.wced.multitask.v1)
- WCED 10 pct [ibm-research/biomed.rna.bert.110m.wced.v1](https://huggingface.co/ibm-research/biomed.rna.bert.110m.wced.v1)

For details on how the models were trained, please refer to [the BMFM-RNA preprint](https://arxiv.org/abs/2506.14861).

### Zero-shot inference (CLI)
`biomed-multi-omic` allows for multiple input data types, but the core data object at the heart of the tool is based around the adata object. Once your dataset is created.

### RNA Inference
To get scRNA embeddings and zero shot cell-type predictions:

```bash
export MY_DATA_FILE=... # path to h5ad file with raw counts and gene symbols
bmfm-targets-run -cn predict input_file=$MY_DATA_FILE working_dir=/tmp checkpoint=ibm-research/biomed.rna.bert.110m.wced.multitask.v1
```

For more details tutorial see RNA tutoirals.

### DNA Inference

To perform DNA zero-shot:

```bash
export MY_DATA_FILE=... # path to your DNA data file
bmfm-targets-run -cn predict input_file=$MY_DATA_FILE working_dir=/tmp checkpoint=ibm-research/biomed.dna.snp.modernbert.113m.v1
```

For more details tutorial see DNA tutoirals.

### Zero-shot inference (programmatically)
To run inference programmatically, you can see a zero-shot example in this [notebook](tutorials/1_zero_shot_cell_type_anno.ipynb).

Note to use the notebook you will need to install the `notebook` optional dependencies (see [Installation](#2-installation))

## Citation

To cite the tool for both RNA and DNA, please cite both the following articles:
```
@misc{dandala2025bmfmrnaopenframeworkbuilding,
      title={BMFM-RNA: An Open Framework for Building and Evaluating Transcriptomic Foundation Models},
      author={Bharath Dandala and Michael M. Danziger and Ella Barkan and Tanwi Biswas and Viatcheslav Gurev and Jianying Hu and Matthew Madgwick and Akira Koseki and Tal Kozlovski and Michal Rosen-Zvi and Yishai Shimoni and Ching-Huei Tsou},
      year={2025},
      eprint={2506.14861},
      archivePrefix={arXiv},
      primaryClass={q-bio.GN},
      url={https://arxiv.org/abs/2506.14861},
}

@misc{li2025bmfmdnasnpawarednafoundation,
      title={BMFM-DNA: A SNP-aware DNA foundation model to capture variant effects}, 
      author={Hongyang Li and Sanjoy Dey and Bum Chul Kwon and Michael Danziger and Michal Rosen-Tzvi and Jianying Hu and James Kozloski and Ching-Huei Tsou and Bharath Dandala and Pablo Meyer},
      year={2025},
      eprint={2507.05265},
      archivePrefix={arXiv},
      primaryClass={q-bio.GN},
      url={https://arxiv.org/abs/2507.05265}, 
}
```
